sdk/go/common/util/gitutil/git_auth.go

```diff
@@ -0,0 +1,329 @@
+// authentication methods, using a set of provided public keys and an ambient SSH agent if
+// available.
+type DefaultSSHAuth struct {
+	user string
+
+	signers        []ssh.Signer
+	sshAgentBroker sshAgentBroker
+
+	gitssh.HostKeyCallbackHelper
+}
+
+var _ AuthMethod = (*DefaultSSHAuth)(nil)
+
+func NewDefaultSSHAuth(user string) DefaultSSHAuth {
+	return DefaultSSHAuth{user: user}
+}
+
+func (s *DefaultSSHAuth) AddIdentityFiles(host string, settings sshUserSettings, passphrase string) error {
+	path := identityFiles(settings, host)
+
+	for _, key := range path {
+		signer, err := parseKey(keyConfig{path: key, passphrase: passphrase})
+		if err != nil {
+			return err
+		}
+		if signer != nil {
+			logging.V(9).Infof("[gitutils] Adding SSH identity file %q", key)
+			s.signers = append(s.signers, signer)
+		}
+	}
+
+	return nil
+}
+
+func (s *DefaultSSHAuth) AddKeyFiles(files ...string) error {
+	for _, file := range files {
+		signer, err := parseKey(keyConfig{path: file, failOnNotExists: true})
+		if err != nil {
+			return err
+		}
+		if signer != nil {
+			logging.V(9).Infof("[gitutils] Adding SSH public key file %q", file)
+			s.signers = append(s.signers, signer)
+		}
+	}
+	return nil
+}
+
+func (s *DefaultSSHAuth) AddKeyFilesWithPassphrase(files []string, passphrase string) error {
+	for _, file := range files {
+		signer, err := parseKey(keyConfig{
+			path:             file,
+			passphrase:       passphrase,
+			failOnNotExists:  true,
+			failOnPassphrase: passphrase != "",
+		})
+		if err != nil {
+			return err
+		}
+		if signer != nil {
+			logging.V(9).Infof("[gitutils] Adding SSH public key file with passphrase %q", file)
+			s.signers = append(s.signers, signer)
+		}
+	}
+	return nil
+}
+
+func (s *DefaultSSHAuth) AddKeyPem(pemData ...[]byte) error {
+	for _, data := range pemData {
+		signer, err := parseKey(keyConfig{data: data})
+		if err != nil {
+			return err
+		}
+		if signer != nil {
+			logging.V(9).Infof("[gitutils] Adding private key PEM")
+			s.signers = append(s.signers, signer)
+		}
+	}
+	return nil
+}
+
+func (s *DefaultSSHAuth) AddKeyPemWithPassphrase(pemData [][]byte, passphrase string) error {
+	for _, data := range pemData {
+		signer, err := parseKey(keyConfig{
+			data:             data,
+			passphrase:       passphrase,
+			failOnPassphrase: true,
+		})
+		if err != nil {
+			return err
+		}
+		if signer != nil {
+			s.signers = append(s.signers, signer)
+		}
+	}
+	return nil
+}
+
+func (s *DefaultSSHAuth) AddSSHAgentBroker(broker sshAgentBroker) error {
+	if s.sshAgentBroker != nil {
+		return errors.New("SSH agent already set")
+	}
+
+	s.sshAgentBroker = broker
+	return nil
+}
+
+type keyConfig struct {
+	path             string
+	data             []byte
+	passphrase       string
+	failOnNotExists  bool
+	failOnPassphrase bool
+}
+
+var DefaultIdentityFiles = []string{"~/.ssh/id_rsa", "~/.ssh/id_dsa", "~/.ssh/id_ecdsa", "~/.ssh/id_ed25519"}
```

Comment:
Looks like this is missing some defaults.

```
GIT_SSH_COMMAND="ssh -v" git clone ssh://...
...
debug1: Will attempt key: /Users/blampe/.ssh/id_rsa RSA SHA256:6BYtFLi+T5uPqqvtqcBiFoUD+dJClE/DygeqKp6eWf4
debug1: Will attempt key: /Users/blampe/.ssh/id_ecdsa
debug1: Will attempt key: /Users/blampe/.ssh/id_ecdsa_sk
debug1: Will attempt key: /Users/blampe/.ssh/id_ed25519 ED25519 SHA256:yc1vKT+VrKDE0e4ZE9bzI0tLk0r82XmsM4tiIHeMCCw
debug1: Will attempt key: /Users/blampe/.ssh/id_ed25519_sk
debug1: Will attempt key: /Users/blampe/.ssh/id_xmss
debug1: Will attempt key: /Users/blampe/.ssh/id_dsa
```

Reply:
It looks like the _sk variants require an interactive yubikey, so I'm leery about adding those in a PR that's a bug fix - we currently don't handle interactive auth "well" here and I don't expect to fix it in this PR. This comment suggests id_dsa and id_xmss should not be included as they're unhandled:

https://github.com/charmbracelet/wishlist/blob/cc7a23ba71506a7f7a7069077836895f6916affa/client_auth.go#L241-L249

That leaves us with id_rsa, id_ecdsa, and id_ed25519. Does that make sense to you?